### PR TITLE
Add PCA-based compression operator

### PR DESCRIPTION
**Pull request by fzimmermann89**
Rel #180

### CODE REVIEW

**Comment by fzimmermann89** [comment]
this looks ugly...

**Comment by fzimmermann89** [comment]
So, coil compression as an operator? Or some function?

**Comment by ckolbPTB** [comment]
Would it make sense to have a PCA operator which is then used by a coil-compression function?

**Comment by fzimmermann89** [comment]
One of the goals is to be able to first construct the compression matrix, then apply it to different data.

So we could move the application to kdata into a function

**Comment by fzimmermann89** [comment]
We should consider how to deal with Data object in operators where it makes sense to apply them to kdata or Idata as well as to tensors.

Either we use some dispatch library, stdlib single dispatch, manual overloads and isinstance or mixing/traits, some optimal subclass manipulation in linear operator (similar to the autograd stuff)

**Comment by fzimmermann89** [comment]
pre-commi.ci autofix

File: src/mrpro/operators/_CoilCompression.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    def __init__(self, data: Data | torch.Tensor, n_components, separate_dims: tuple[int, ...] | None = None):
+        """Construct a PCA based soft coil compression operator.
+
+        Parameters
+        ----------
+        data
+            Data to used to find the principal components.
+        n_components
+            Number of principal components to keep.
+        separate_dim, optional
```
Comment (fzimmermann89) [inline src/mrpro/operators/_CoilCompression.py]:
```suggestion
        separate_dims
```

File: src/mrpro/operators/_DataOperator.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    """
+
+    def __init__(self, operator: Operator[torch.Tensor, tuple[torch.Tensor,]], return_type: type[Data] | None = None):
+        """Initialize the DataOperator.
+
+        Parameters
+        ----------
+        Operator
+            The operator to apply
+        returntype
```
Comment (fzimmermann89) [inline src/mrpro/operators/_DataOperator.py]:
```suggestion
        operator
            The operator to wrap
        return_type
```

File: src/mrpro/operators/_DataOperator.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import dataclasses
+from copy import deepcopy
+
+import torch
+
+from mrpro.data import Data
+from mrpro.operators import Operator
+
+
+class DataOperator(torch.nn.Module):
```
Comment (fzimmermann89) [inline src/mrpro/operators/_DataOperator.py]: Spit out of this pr

File: src/mrpro/operators/_CoilCompression.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import torch
+
+from mrpro.data import Data
+from mrpro.operators import LinearOperator
+
+
+class CoilCompression(LinearOperator):
+    """PCA based soft coil compression operator."""
+
+    def __init__(self, data: Data | torch.Tensor, n_components, separate_dims: tuple[int, ...] | None = None):
```
Comment (fzimmermann89) [inline src/mrpro/operators/_CoilCompression.py]:
```suggestion
    def __init__(self, data: Data | torch.Tensor, n_components:int, separate_dims: tuple[int, ...] | None = None):
```

File: src/mrpro/operators/_CoilCompression.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import einops
+import numpy as np
+import torch
+
+from mrpro.data import Data
+from mrpro.operators import LinearOperator
+
+
+class CoilCompression(LinearOperator):
```
Comment (fzimmermann89) [inline src/mrpro/operators/_CoilCompression.py]: Rename to PCACompressionOp

**Comment by fzimmermann89** [comment]
Does anybody want to take this PR?

The signature should be changed, instead of separate_dims as an argument and only working on dim -4, there should be a dim argument which dimension to compress, defaulting to -4.
And a shared_dims (tuple of int or ellipsis) with the opposite meaning of separate dims. A single compression matrix should be constructed for these dimensions, with ellipsis (default) meaning all dimensions. For dimensions not specified as shared, normal natxhing shall be used.

The data operator part should just be removed here.

**Comment by fzimmermann89** [comment]
@ckolbPTB are you interested?

**Comment by ckolbPTB** [comment]
To me separate dims seems to more intuitive parameter and easier to define if the shape of data used for calculating the compression matrix is different to the input data to forward.

**Comment by ckolbPTB** [comment]
I would then add a `compress_coils` algorithm in a separate PR

**Comment by fzimmermann89** [comment]
> To me separate dims seems to more intuitive parameter and easier to define if the shape of data used for calculating the compression matrix is different to the input data to forward.

yes, the implementation would be easier, but using it would be a bit inconsistent, as we assume all leading dimensions are always batch dimensions. 
otherwise adding another leading dimensions needs a code change to batch along that dimensions. this is not what we want.

This is why I deliberately suggested having the inverse, a joint_dim argument -- and why I tried to find somebody else to tackle this ;)

**Comment by ckolbPTB** [comment]
> yes, the implementation would be easier, but using it would be a bit inconsistent, as we assume all leading dimensions are always batch dimensions.

I did not mean the implementation but the usage. One example woud be PCA-based compression for qMRI. Let's assume we have one dimension a which is supposed to be treated separately. We would calculate N signal curves in the shape (a,t,N) and then calculate the compression matrix for this. This compression matrix would then be applied to the data of shape (a,t,z,y,x). If I used joint_dim then I would have to specific joint_dim=-1 for the init but then change it when applying to the data. If I used separate_dims it would be 0 for both. 

Also for coil compression I would like to define the separate_dim as the slice dimension and not care if I have additional leading dimensions due to dynamics or echos or something like this.

**Comment by fzimmermann89** [comment]
If I am not mistaken we have to specify four dimensions:
- the dimension in which to compress (coils or time)  when creating the compression 
- dimension to compress when applying the data
-  dimensions to batch over vs dimensions to consider as samples on the data matrix when creating the compression 
- how to broadcast the result at application 

 
Either we allow to specify all of them or certain applications require either unsqueezes or reaaranges.

We could just keep the separate&dim logic for know, although it is kind of a foot gun that adding a batch dimension in front of the data will make this by default a shared dimension.

Maybe at some point er can come up with better definitions here what do to.

**Comment by fzimmermann89** [comment]
>
> Also for coil compression I would like to define the separate_dim as the slice dimension and not care if I have additional leading dimensions due to dynamics or echos or something like this. 

This breaks both pytorch and somewhat mrpro conventions: leading dimensions are considered batch dimensions, thus should be treated as separate.
 If you want to share compression along dynamics, for example, you should specify that. If you then use batches of different subjects, it should treat the new 0tj dim as a separate batch dim. So the logic is exactly the opposite.

**Comment by fzimmermann89** [comment]
If I am not mistaken we have to specify three dimensions:
- the dimension in which to compress (coils or time)  when creating the compression 
- dimension to compress when applying the data
-  dimensions to batch over vs dimensions to consider as samples on the data matrix when creating the compression (this is where we disagree) - do you want to use the same compression for different slices, subjects etc or the same?
 
Either we allow to specify all three, or certain applications require either unsqueezes or reaaranges.

I used the code to construct one coil compression operator based on one sample and applied it to differently shaped inputs with different batch dimensions
So it would be creation_dim=coil, joint_dim=z y x, application_dim=coil. At application this should now broadcast in the leading dimensions of the data and the same compression is applied for rotation,slice,....

In a qmri example, you would want to create the matrix with (t,N)-shaped samples, and compress in time. You want to apply it to data 
( t, slice, rotation, coil,zyx)
Here the creation dim is 0, the application dim is 0 and the there is no joint_dim.

If the data were 
( Subject, t, slice, rotation, coil,zyx) you would have to specify an application dimension or create the matrix with (subject or 1,t,N) samples.

Specifying separate dimensions instead of joint dimensions would not help here.

**Comment by fzimmermann89** [comment]
Also, thinking about it, maybe the differences between use in qmri and coil compression are greater than the common parts. And itigjt ne easier to just have to things that do a bit of (different) reshaping before applying a PCA ..?

**Comment by ckolbPTB** [comment]
> We could just keep the separate&dim logic for know, although it is kind of a foot gun that adding a batch dimension in front of the data will make this by default a shared dimension.

But isn't this the definition of a batch dimension?

**Comment by fzimmermann89** [comment]
> > We could just keep the separate&dim logic for know, although it is kind of a foot gun that adding a batch dimension in front of the data will make this by default a shared dimension.
> 
> But isn't this the definition of a batch dimension? 

It would usually be separate dimension, i.e. having a batch dimension is equivalent to having a for loop over the different batch samples. So no joint operations connecting the batch samples (beside batch norm as THE special case)
L

**Comment by fzimmermann89** [comment]
Example:  having a batch of three subjects should do separate coil compressions for each of the subjects.

**Comment by ckolbPTB** [comment]
@fzimmermann89 could you have a look at this please?

File: src/mrpro/operators/PCACompressionOp.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        The operator carries out an SVD followed by a threshold of the n_components largest values along the last
+        dimension of a data with shape (*other, joint_dim, compression_dim). A single SVD is carried out for everything
+        along joint_dim. Other are batch dimensions.
+
+        Consider combining this operator with RearrangeOp to make sure the data is in the correct shape before applying.
+
+        Parameters
+        ----------
+        data
+            Data to be used to find the principal components of shape (*other, joint_dim, compression_dim)
```
Comment (schuenke) [inline src/mrpro/operators/PCACompressionOp.py]: Maybe change to:
`Data of shape (*other, joint_dim, compression_dim) to be used to find the principal components. `
Else sounds like the components have that shape...

File: src/mrpro/operators/PCACompressionOp.py
Comment (schuenke) [inline src/mrpro/operators/PCACompressionOp.py:1]: Maybe we want to add the option to determine n_components automatically from the data? For an old project I implemented this using one approach and looked up two other ones, that I didn't implement yet. Not sure if we want this option (n_components could then maybe be of type int or Literal(implemented methods)), but I will paste my old code/references here just in case:

```python
def det_malinowski_factor(eigenvalues: np.ndarray) -> int:
    """
    Determines minimum number of principal components according to: Malinowski, E. R. Determination of the number of
    factors and the experimental error in a data matrix. Anal. Chem. 49, 612–617 (1977).
    """
    ind = np.zeros([eigenvalues.shape[0]-1, 1])
    for n in range(ind.shape[0]):
        ind[n] = np.sqrt(np.sum(eigenvalues[n:]/(eigenvalues.shape[0]-(n+1)))) / (eigenvalues.shape[0]-(n+1))**2

    return int(np.argmin(ind))


def det_nelson_factor(eigenvalues: np.ndarray) -> int:
    """
    Determines minimum number of principal components according to: Nelson, L.R. (2005). Some observations on the
    scree test, and on coefficient alpha. Thai Journal of Educational Research and Measurement: 3(1), 1-17.
    """
    # Todo: implement function :)
    pass


def det_median_noise_factor(eigenvalues: np.ndarray) -> int:
    """
    Determines minimum number of principal components according to: Manjón, J.V. et al. (2015). MRI noise estimation and
    denoising using non-local PCA. Medical Image Analysis 22 (2015) 35–47
    """
    # Todo: implement function :)
    pass

```

**Review by schuenke — APPROVED** [review]

File: src/mrpro/operators/PCACompressionOp.py
Comment (ckolbPTB) [reply src/mrpro/operators/PCACompressionOp.py:1]: new issue #479

**Comment by ckolbPTB** [comment]
@fzimmermann89 could you merge this or is anything still missing?

**Comment by fzimmermann89** [comment]
I will check later and merge it.

**Comment by fzimmermann89** [comment]
pre-commit.ci autofix

**Comment by fzimmermann89** [comment]
Does anybody know if it is possible to make RearrangeOp a link in the documentation?

**Comment by fzimmermann89** [comment]
> Does anybody know if it is possible to make RearrangeOp a link in the documentation?

? @lrlunin or @ckolbPTB

**Comment by ckolbPTB** [comment]
> Does anybody know if it is possible to make RearrangeOp a link in the documentation?

I don't think I understand the question....

**Comment by fzimmermann89** [comment]
In function signatures etc, class names are linked in the documentation to the respective page.

I was trying to make the reference to `RearrangeOp` in the docstring a clickable link in the documentation -- but failed.

**Comment by fzimmermann89** [comment]
image
vs 
image
